### PR TITLE
don't blindly jump releases

### PR DESCRIPTION
**Pull request by timkuijsten**
I propose to stop tracking the moving target Debian Stable because this renders a lot of demos broken at the time Debian releases a new stable. A situation that can last for months depending on the schedule and availability of the developer of a particular demo. During this time the demo can no longer be showed to or used by third parties that might be interested in the project.

Instead, track the latest stable that is known to work and as soon as Debian releases a new stable test all demos with it and coordinate this with the individual owners of each demo before bumping the base image.

For now track Debian Stretch because that is the latest stable that is known to work and thus supported by all of us. Since Buster has been released this summer, work should be initiated to make sure all demos work with Buster so that we can move the base image to it.

Maybe we can create an extra image during the transition to the new stable so that we can all test and sort of flag when our own demos are ready. Maybe call it "nextbase"? I'm not sure how much work an extra image like this would be in the current infrastructure.

### CODE REVIEW

**Comment by tvrancken** [comment]
I think this is a good idea. So instead of following the "stable" tag, which is a moving target, we follow a specific stable version (e.g. Stretch) which remains the same indefinitely (and thus does not break our demos). Then before we move to the next stable (when it is released) we first test all our programs/demos and implement fixes accordingly.

**Comment by amarsman** [comment]
I agree. Let's fix the master branch to a specific Debian stable version, create a level branch that can be used to fix issues and only when everything works again merge to master. In the devel branch we can also update the Debian stable version, when that works for everything merge to master. So master should always work, devel might not but when we reach a stable branch again merge to master.

I will work on CI to be able to support this workflow but we can already start working this way.

**Review by amarsman — APPROVED** [review]

**Comment by vanrein** [comment]
It was a very deliberate choice to pick a moving target, and also explained back then.  The stability of a demo follows from the latest tags that each build target got in Git.  These tags are coupled to builds of a binary at the Docker Hub.  Binary is the stable form, source aims for a moving target.  I see nothing strange in that.

Remember when the TLS Pool was a disaster to replicate?  The cause was that its development environment got stuck in an old distribution, for completely trivial but still blocking reasons.  This sort of local issues are easily resolved, and the moving target should motivate that.  It is the best way to standardise without locking ourselves into a moment in time and the need to make big changes in many places at the same time.  If we changed from, say, `stable` to `jessie`, we would be giving rise to this sort of trouble.

I do agree that long-lasting inability to build what we post is undesirable, even if it just for sharing code among ourselves.  It is a sign that we are not actively using all/the Docker Demos ourselves.  To that end, the proposal of building it the `demo-*` images on a daily basis, as proposed by Aschwin, adds true value.  The fix needed is usually trivial for the developer who made the demo, and this kind of attention will be needed sooner or later.

What this issue is raising, as far as I'm concerned, is that the Docker Demos are not for end user consumption; they should be pointed to the Docker Hub instead, or provided other binaries.  That hardly sounds like a problem to me.  End users may still be invited to the source of course, but there nobody should be surprised about occasionally ramblings due to incompatible shifts in underlying dependencies.

In short, I do not believe that moving away from Debian Stable is helpful, actually quite the contrary.  Monitoring the proper builds on a daily basis is a better idea, I think.

**Comment by vanrein** [comment]
Re. the patch: This is almost certainly not the place to make this change; it works now and fails later.

All Docker Demos inherit from this and would have to change at the same time.  If we went in this direction, which I strongly discourage above, we would better use separate `base` packages and allow each Demo to switch at its earliest convenience.  But will it continue to work together, especially when taking code from a  build image and loading it into a demo image?

You guessed it, we're introducing another coordination problem.  This is down the rabbit hole, I'm afraid.

If you really insist on a version specific to a static Debian distro, an alternative idea might be to use a branch where pretty much this change is made.  That too would have the problem of shared inheritance, though.

I think you are shifting muck without actually getting rid of it :) and that's why I believe the muck-spotting proposal of Aschwin is the better way to go, and avoid the soon-to-come version splitting proposed here.

**Comment by amarsman** [comment]
> On 15 Nov, 2019, at 19:15 , vanrein <notifications@github.com> wrote:

Hi,

Thanks for the extensive feedback.

> It was a very deliberate choice to pick a moving target, and also explained back then. The stability of a demo follows from the latest tags that each build target got in Git. These tags are coupled to builds of a binary at the Docker Hub. Binary is the stable form, source aims for a moving target. I see nothing strange in that.

It’s clear if you are a part of arpa2 but it has a big draw back: you only get a first impression once.

The existence of master for a stable version which always works and gives the good first impression: you git clone it, build it and it just works.

For development you have a level branch, which aims to be stable but doesn’t have to be for a “short while”. When all builds and tests succeed for development it will be merged to master.

In 2010 Vincent Driessen described this what was called later gitflow and a lot of tools have support for it and it’s very well documented, no need to re-invent the wheel. [1]

> Remember when the TLS Pool was a disaster to replicate? The cause was that its development environment got stuck in an old distribution, for completely trivial but still blocking reasons. This sort of local issues are easily resolved, and the moving target should motivate that. It is the best way to standardise without locking ourselves into a moment in time and the need to make big changes in many places at the same time. If we changed from, say, stable to jessie, we would be giving rise to this sort of trouble.

It’s not only that things break when there is a major update of Debian stable, it also breaks if one of our dependencies is changed in an incompatible way. When that happens we fix it in a (branch of) devel, test it and if successful merge to master and our master is always working and gets updated quickly when something fails.

> I do agree that long-lasting inability to build what we post is undesirable, even if it just for sharing code among ourselves. It is a sign that we are not actively using all/the Docker Demos ourselves. To that end, the proposal of building it the demo-* images on a daily basis, as proposed by Aschwin, adds true value. The fix needed is usually trivial for the developer who made the demo, and this kind of attention will be needed sooner or later.

If we want to be taken serious as a project we needs to respond to broken builds and failing tests.

I propose a CI system that builds at least master and devel on a daily basis. Master should always succeed, devel can fail, we notice that within 24 hours and it will be fixed with priority. When the CI system indicates that level is successful it will be merged to master and all is well.

To get a quick feedback loop also your own branch can be build with the CI system, when your branch is stable you merge it to devel and we take it from there.

> What this issue is raising, as far as I'm concerned, is that the Docker Demos are not for end user consumption; they should be pointed to the Docker Hub instead, or provided other binaries.  

I strongly disagree: the docker demos are an easy way to build (a part of) arpa2 and it’s the most convenient way for an interested developer to look into arpa2.

> That hardly sounds like a problem to me. End users may still be invited to the source of course, but there nobody should be surprised about occasionally ramblings due to incompatible shifts in underlying dependencies.

If we want to be successful as a project the first impression is crucial: if it’s not easy to build and get started, people won’t join.

> In short, I do not believe that moving away from Debian Stable is helpful, actually quite the contrary. Monitoring the proper builds on a daily basis is a better idea, I think.

As will be clear by now, I disagree. The way I described it works in successful commercial companies and open source projects and I like a process that has proven itself for many years already.

Have a nice weekend,

Aschwin

[1] https://nvie.com/posts/a-successful-git-branching-model/

**Comment by amarsman** [comment]
> On 15 Nov, 2019, at 19:27 , vanrein <notifications@github.com> wrote:
> 
> Re. the patch: This is almost certainly not the place to make this change; it works now and fails later.

I think this -is- the correct place, in master it’s fixed, when there is a new Debian stable we change it in devel, make it work again and merge to master.

> All Docker Demos inherit from this and would have to change at the same time. 

This is what you have at the moment but not in a controlled way.

> If we went in this direction, which I strongly discourage above, we would better use separate base packages and allow each Demo to switch at its earliest convenience. But will it continue to work together, especially when taking code from a build image and loading it into a demo image?

This is not the way to go, we can have some settle time after a version bump but if we can’t get into a working state again in a couple of weeks we have really failed as a project.

> You guessed it, we're introducing another coordination problem. This is down the rabbit hole, I'm afraid.

I won’t repeat myself.

> If you really insist on a version specific to a static Debian distro, an alternative idea might be to use a branch where pretty much this change is made. That too would have the problem of shared inheritance, though.
> 
> I think you are shifting muck without actually getting rid of it :) and that's why I believe the muck-spotting proposal of Aschwin is the better way to go, and avoid the soon-to-come version splitting proposed here.

Everything is already said, let’s hear the opinion of other team members.

Have a nice weekend,

Aschwin

**Comment by vanrein** [comment]
Sounds like we have different purposes in mind for this repository.  It's useful to clarify that during an upcoming meeting.  After resolving that it should be straightforward enough to find implementation details and ways of co-existence for all the goals that we decide to continue/adopt.

**Comment by vanrein** [comment]
The purpose for which I had made the Docker Demos is to use among ourselves.  The sharing of state is in the service of sharing of image disk space through Docker layers.  It may be too strict if we stick to one distribution, at the very least another branch could be created to use another Debian distro.

Meanwhile, people have started to think of this project as aimed at end users.  This was indeed a vague idea behind it, once, but it is clear that this is too much in swing for that to be solid and reliable.  Docker also has shown to be less usable (less reliable) for deployment, as far as I'm concerned.

For this reason, I've started two projects for building our software:

  * [mkhere](https://github.com/arpa2/mkhere) was built on the observation that it would be useful to point at a directory structure as a build file system.  To that end, the `mkhere` project is installed in the root of this build file system, and symbolic links named `mkhere` might reference that directory.  Projects might be built with statements like `mkhere/tlspool.sh have depend build` and they may be exported as simple archives, possibly with libraries used from the underlying OS.  Packages installed on the underlying OS may also be exported as simple archives.
  * [mkroot](https://github.com/arpa2/mkroot) was built on the observation that most deployment scenarios require a root filesystem (directory) and additional resource accessibility declarations.  Examples are the OCI Bundles for OpenContainers/RunC, PXE/SYSLINUX images to be loaded as `initrd` by a separately retrieved `vmlinuz` kernel.  The intention is make deployable container images that still need some personalisation, and have places to enable just that.

The last two projects can easily become subject to the same kind of referencing-externally-instable-software problem.  Since it is not targeted at development however, they may be more geared towards release engineering.  Meanwhile, the Docker Demos could be useful development tooling among ourselves, and they may be a rough version prefacing releases in `mkhere` and `mkroot` projects.

We should ask ourselves how release engineering for `mkhere` and `mkroot` should be done, and whether sufficient options remain for personalisation.

**Comment by amarsman** [comment]
> On 25 Nov, 2019, at 11:36 , vanrein <notifications@github.com> wrote:
> 
> The purpose for which I had made the Docker Demos is to use among ourselves. The sharing of state is in the service of sharing of image disk space through Docker layers. It may be too strict if we stick to one distribution, at the very least another branch could be created to use another Debian distro.
> 
> Meanwhile, people have started to think of this project as aimed at end users. This was indeed a vague idea behind it, once, but it is clear that this is too much in swing for that to be solid and reliable. Docker als has shown to be less usable (less reliable) for deployment, as far as I'm concerned.

I’m sure Docker is not perfect but it works well enough for this phase of the project: easily be able to build and extend an image. It’s huge step forward with how it was done 3 years ago.

> For this reason, I've started two projects for building our software:

For me Docker is good enough. I don’t like to have to learn another way of working at the moment, it’s easy to pull the arpa2/docker-demo repo and build an image you are interested in.

As far as I’m concerned this project needs new components that work together to get closer to the end goal. Spending time on yet another way of building is not on my priority list.

If you have the time and are limited in the pace of development with Docker please go ahead but I don’t see the use case at this point in time.

> 	• mkhere was built on the observation that it would be useful to point at a directory structure as a build file system. To that end, the mkhere project is installed in the root of this build file system, and symbolic links named mkhere might reference that directory. Projects might be built with statements like mkhere/tlspool have depend build and they may be exported as simple archives, possibly with libraries used from the underlying OS. Packages installed on the underlying OS may also be exported as simple archives.
> 	• mkroot was built on the observation that most deployment scenarios require a root filesystem (directory) and additional resource accessibility declarations. Examples are the OCI Bundles for OpenContainers/RunC, PXE/SYSLINUX images to be loaded as initrd by a separately retrieved vmlinuz kernel. The intention is make deployable container images that still need some personalisation, and have places to enable just that.
> The last two projects can easily become subject to the same kind of referencing-externally-instable-software problem. Since it is not targeted at development however, they may be more geared towards release engineering. Meanwhile, the Docker Demos could be useful development tooling among ourselves, and they may be a rough version prefacing releases in mkhere and mkroot projects.

Docker is well known, no need to educate users ourselves. For final deployment you probably want to have the method that your distribution or OS has already.

> We should ask ourselves how release engineering for mkhere and mkroot should be done, and whether sufficient options remain for personalisation.

Any opinions?

Best regards,

Aschwin

**Comment by timkuijsten** [comment]
>  The cause was that its development environment got stuck in an old distribution, for completely trivial but still blocking reasons. This sort of local issues are easily resolved, and the moving target should motivate that. It is the best way to standardise without locking ourselves into a moment in time and the need to make big changes in many places at the same time. If we changed from, say, stable to jessie, we would be giving rise to this sort of trouble.

I quite strongly agree with this.

> I will work on CI to be able to support this workflow but we can already start working this way.

Also a very good idea, independent of this issue about following `stable` or `latest-stable-supported-by-us`.

> The purpose for which I had made the Docker Demos is to use among ourselves.

When I opened this ticket I didn't think about Docker Hub and the idea that users should pull the images from Docker Hub instead of building the image themselves. Since I have used Docker Hub only once and it was not very stable at the time and I forgot about the exact steps on how to update/build the image I now have dropped the requirement on it and simply [instruct](/arpa2/docker-demo/commit/7ad8f18a2751011839b88c3b8dca70621cf527c7) users to use `docker build` instead of `docker pull` like I already always did with demo-acl. As far as I'm concerned we drop the extra steps and cognitive load of making our images available on Docker Hub.
